### PR TITLE
firebase serv add option to override the public directory

### PR DESCRIPTION
**Pull request by benyitzhaki**
add the option to override the public directory that is been served while using firebase serv. the directory can been overridden using a flag --public-hosting

this is regarding the following issue:
https://github.com/firebase/firebase-tools/issues/234

note: the public flag was already taken

### CODE REVIEW

**Comment by benyitzhaki** [comment]
I signed it!

File: commands/serve.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ module.exports = new Command('serve')
   .option('-o, --host <host>', 'the host on which to listen (default: localhost)', 'localhost')
   .option('--only <targets>', 'only serve specified targets')
   .option('--except <targets>', 'serve all except specified targets')
+  .option('--hosting-public <targets>', 'override public directory')
```
Comment (laurenzlong) [inline commands/serve.js]: I think "hosting-dir" might be a better flag name. Also should replace "\<targets\>" with "\<directory\>". @mbleigh what do you think?

Comment (laurenzlong) [inline commands/serve.js]: Would also recommend changing description to: 'override hosting directory for serving'

Comment (benyitzhaki) [inline commands/serve.js]: @laurenzlong I agree with your change requests and made them

**Comment by mbleigh** [comment]
Thanks for the PR! We have a pretty strict internal process for changes to our commands so I can't merge this until we've done some process around whether we think this is a good API change (and the right name for things). Just wanted to give you a heads up that this may take some time to land, but I'll start the discussion internally.

**Comment by eskan** [comment]
hi,
this is a wanted future as you develop with polymer you need to serve client sources directory instead of the build directory you should deploy.
We can use **polymer serve** but if you use firebase reserved url (/__/) you need to serve src with firebase tools cli.
So please consider this enhancement.
Cheers

**Comment by danizord** [comment]
Same as @eskan here. I'm also using Polymer with `firebase serve` and I want the development server to serve from my root project directory instead of build directory, which is only used for deployment.

I'm surprised the `--public` option is currently supported by `firebase deploy`, where I think it's rarely used, but it is not supported by `firebase serve`, where it is often a common need to change the root directory 😄

**Comment by samtstern** [comment]
Is this still in progress?  Is it solved by `firebase.json`?

**Comment by bkendall** [comment]
@samtstern I don't think it's necessarily solved, but I don't think this is going to be something we add in the near future.

It makes sense to me that `firebase deploy` takes an override, because that's where you probably want to have your minified or smallest version of the files to be deployed. `firebase serve` should definitely serve as a tool for development, so having that set to a standard build path (where something like webpack would continually build to) makes sense as well. I'm struggling to see why I would want yet another way to override `serve`. If "being able to test both the minified and development version" is in the cards, I'd argue that the build folder should always be the same, and other scripts take care of putting the right content in that folder.

Because this issue is so old and isn't actively being considered, I'm going to close it. If anyone has very strong motivations to make a case for supporting this behavior further, please comment and/or make a new issue, and we can have that discussion.

**Comment by danizord** [comment]
@bkendall in some development environments such as Polymer, there's no "development build", since the source files are already runnable by browser. We use webpack only to generate production bundles.

**Comment by eskan** [comment]
I still don't understand why an enhancement ask by users, it's not considered. And all this is to implemented full google stack development. The request is simple we want to serve **for development** a different directory than the production one, it's not a big deal ?!
Because it's such a pain to bundle client sources and we loose tools like browser sync.
You need to know as a commercial enterprise, that kind of feature that is a developer brake have made me move to graphql even if your environment is so efficient.
Regards,

**Comment by bkendall** [comment]
@danizord gotcha. To make sure I understand your flow, you have something like

```
project/
  firebase.json
  src/
    index.html
  build/
    [generated files]
```

And your `firebase.json` has the following as well (short form, for brevity):

```
hosting.public = "build"
```

So, this said, the deployment flow is fairly obvious: `firebase deploy --only hosting` will take the contents of `project/build/` and deploy them. The local development story is a little less obvious. So if `firebase serve` is used, it will serve the files in `project/build/` with nearly the same behavior as if it was being served from your `firebaseapp.com` domain (via Hosting). But that's not desired; what is desired is that the `project/src/` directory is served as that's actually useful for debugging (and stuff).

I can think of some reasons why having `serve` be more similar to production than development would make sense, but I want to hear from you. Would you mind telling me:

1. Is my story here correct/similar enough to your development flow?
1. How would _you_ do development normally? Do you have a small web server that serves `project/src/` and keeps the files updated?
1. From a "normal" development flow (not using Firebase), what becomes an issue when starting to introduce Firebase things?

I'm trying to dig down a little bit to get to the root of the problem. Maybe the understanding of what `serve` does and should do are not completely aligned, which is what would cause some of this turmoil. Thanks for sticking with me - these answers will go a long way to getting this resolved!

**Pull request closed without merging**